x/evm/keeper/state_transition.go

```diff
@@ -167,23 +168,23 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 	res.Hash = txHash.Hex()
 	logs := k.GetTxLogs(txHash)
 
-	// Commit and switch to committed context
+	// Commit if success
 	if !res.Failed() {
-		commit()
-	}
+		k.CommitCachedContexts()
 
-	k.EndCachedContext()
-
-	// Logs needs to be ignored when tx is reverted
-	// Set the log and bloom filter only when the tx is NOT REVERTED
-	if !res.Failed() {
+		// Logs needs to be ignored when tx is reverted
+		// Set the log and bloom filter only when the tx is NOT REVERTED
 		res.Logs = types.NewLogsFromEth(logs)
 		// Update block bloom filter in the original context because blockbloom is set in EndBlock
 		bloom := k.GetBlockBloomTransient()
 		bloom.Or(bloom, big.NewInt(0).SetBytes(ethtypes.LogsBloom(logs)))
 		k.SetBlockBloomTransient(bloom)
 	}
 
+	if !k.ctxStack.IsClean() {
```

Comment:
In case of res.Failed, won't the cache be dirty since you are not commiting?

Reply:
If failed, the evm internal should already called the `RevertToSnapshot`.

Reply:
ah okay, maybe in that case you can always commit so that it would be more clear

Reply:
hmm or just put a comment

Reply:
> ah okay, maybe in that case you can always commit so that it would be more clear

yeah, sounds good.